Python/symtable.c

```diff
@@ -534,6 +579,30 @@ analyze_name(PySTEntryObject *ste, PyObject *scopes, PyObject *name, long flags,
             return 0;
         if (PySet_Discard(global, name) < 0)
             return 0;
+        if (flags & DEF_TYPE_PARAM) {
+            if (PySet_Add(typeparams, name) < 0)
+                return 0;
+        }
+        else {
+            if (PySet_Discard(typeparams, name) < 0)
+                return 0;
+        }
+        return 1;
+    }
+    // If we were passed class_symbols (i.e., we're in an ste_can_see_class_scope scope)
+    // and the bound name is in that set, then the name is potentially bound both by
+    // the immediately enclosing class namespace, and also by an outer function namespace.
+    // In that case, we want the runtime name resolution to look at only the class
+    // namespace and the globals (not the namespace providing the bound).
+    // Similarly, if the name is explicitly global in the class namespace (through the
+    // global statement), we want to also treat it as a global in this scope.
+    long class_flags = flags_in_symbols(class_symbols, name);
+    if (class_flags & DEF_GLOBAL) {
+        SET_SCOPE(scopes, name, GLOBAL_EXPLICIT);
```

Comment:
Is there a case where you found it necessary to ensure this is `GLOBAL_EXPLICIT` rather than `GLOBAL_IMPLICIT`? I note that below we just use `GLOBAL_IMPLICIT` for the "global statement found in enclosing scope" case, although there's a comment suggesting maybe it should be explicit instead.

Reply:
Yes, in the compiler we now emit LOAD_GLOBAL for GLOBAL_EXPLICIT but LOAD_FROM_DICT_OR_GLOBALS for GLOBAL_IMPLICIT, because for explicit globals, we should not check the class dict.

Reply:
Makes sense. I was tempted to say that we could always use `LOAD_FROM_DICT_OR_GLOBALS`, since if something is an explicit global in the class namespace it can't also be bound there. But of course this is only true statically; since the class namespace is dynamic, we have no idea what might be inserted by a metaclass `prepare`, or monkeypatched in later, including something that conflicts with an explicit `global` statement. So I guess we do need to make this distinction.

Reply:
Yes. It also matters if you inject names in `__prepare__` or with an explicit mutation to `locals()`.